shared/provision/code-page/index.js

```diff
@@ -0,0 +1,479 @@
+      props.currentDeviceAlreadyProvisioned ? oppositeTabMap[tabToShowToNew] : tabToShowToNew
+
+    if (props.currentDeviceType === 'mobile') {
+      return getTabOrOpposite('QR')
+    } else if (props.currentDeviceType === 'desktop') {
+      return props.otherDeviceType === 'desktop' ? getTabOrOpposite('viewText') : getTabOrOpposite('QR')
+    }
+
+    throw new Error('Impossible defaultTab')
+  }
+
+  _tabBackground = () => (this.state.tab === 'QR' ? globalColors.blue2 : globalColors.green)
+
+  render() {
+    let content
+    switch (this.state.tab) {
+      case 'QR':
+        content = <Qr {...this.props} />
+        break
+      case 'viewText':
+        content = <ViewText {...this.props} />
+        break
+      case 'enterText':
+        content = <EnterText {...this.props} />
+        break
+      default:
+        /*::
+      declare var ifFlowErrorsHereItsCauseYouDidntHandleAllTypesAbove: (action: empty) => any
+      ifFlowErrorsHereItsCauseYouDidntHandleAllTypesAbove(this.state.tab);
+      */
+        content = null
+    }
+
+    return (
+      <Box2
+        direction="vertical"
+        fullWidth={true}
+        fullHeight={true}
+        style={collapseStyles([styles.codePageContainer, {backgroundColor: this._tabBackground()}])}
+      >
+        <Box2
+          direction="vertical"
+          fullHeight={true}
+          style={
+            this.props.currentDeviceAlreadyProvisioned
+              ? styles.imageContainerOnLeft
+              : styles.imageContainerOnRight
+          }
+        >
+          <RequireImage
+            src={this.state.tab === 'QR' ? blueBackground : greenBackground}
+            style={
+              this.props.currentDeviceAlreadyProvisioned ? styles.backgroundOnLeft : styles.backgroundOnRight
+            }
+          />
+        </Box2>
+        <BackButton
+          onClick={this.props.onBack}
+          iconColor={globalColors.white}
+          style={styles.backButton}
+          textStyle={styles.backButtonText}
+        />
+        {!!this.props.error && <ErrorBanner error={this.props.error} />}
+        <Box2 direction="vertical" style={styles.container} fullWidth={true}>
+          <Instructions {...this.props} />
+          {content}
+          <SwitchTab {...this.props} selected={this.state.tab} onSelect={tab => this.setState({tab})} />
+        </Box2>
+      </Box2>
+    )
+  }
+}
+
+const ErrorBanner = (props: {error: string}) => (
+  <Box2 direction="vertical" style={styles.errorContainer}>
+    <Text type="Body" style={styles.errorText}>
+      {props.error}
+    </Text>
+  </Box2>
+)
+
+const SwitchTab = (props: {|...Props, selected: Tab, onSelect: Tab => void|}) => {
+  if (props.currentDeviceType === 'desktop' && props.otherDeviceType === 'desktop') {
+    return <Box2 direction="horizontal" />
+  }
+
+  let label
+  let icon
+  let tab
+
+  if (props.selected === 'QR') {
+    label = 'Type secret instead'
+    icon = 'iconfont-arrow-right'
```

Comment:
Looks like the Zeplin design has new icons for these, rather than reusing arrow -- perhaps those are coming later?

Reply:
updated in part 4